### PR TITLE
Fix 3-D config for webgl buffer on iPad Pro & iPad 7th + iOs v13 + Safari 

### LINKED ISSUE #4502: iPad Pro 12.9 (13) and iPad 7th (13) Not Detected Properly using Safari
There appears to be an issue with WebGL detection on larger iPad Pro 12.9" tablets.  I found ticket #280 that addressed this at one point for other iOS devices, but not sure if this is a regression or not.  Additional details can be found [here](https://github.com/jneilliii/OctoPrint-BedLevelVisualizer/issues/180) of the problem.  It seems that if the browser is made smaller the graphing works fine, but once maximized to the full size of the tablet, the error "WebGL is not supported by your browser" is displayed.  The same result happens for this user as well on the example pages [here](https://plot.ly/javascript/3d-surface-plots/).

### PR DESCRIPTION
**Pull request by archmoj**
Fixes #4502 | [Demo](https://codepen.io/MojtabaSamimi/pen/gObVddE?editors=1000)

@plotly/plotly_js 
cc: @jackparmer

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -82,7 +82,7 @@
     "gl-mat4": "^1.2.0",
     "gl-mesh3d": "^2.3.0",
     "gl-plot2d": "^1.4.3",
-    "gl-plot3d": "^2.4.1",
+    "gl-plot3d": "^2.4.2",
```
Comment (archmoj) [inline package.json]: https://github.com/gl-vis/gl-plot3d/pull/27

File: src/plots/gl3d/scene.js [hunk 2]
```diff
@@ -31,6 +30,32 @@ var createAxesOptions = require('./layout/convert');
...
+        tablet: true
+    });
+
+    // handle iPad pro or iPad with iOs 13 using Safari
+    // see https://github.com/plotly/plotly.js/issues/4502
+    if(
+        result === false &&
+        userAgent.indexOf('Macintosh') !== -1 &&
+        userAgent.indexOf('Safari') !== -1 &&
+        navigator.maxTouchPoints > 1
```
Comment (etpinard) [inline src/plots/gl3d/scene.js]: How did you find this solution?

Comment (archmoj) [reply src/plots/gl3d/scene.js]: Inspired by info on https://stackoverflow.com/questions/58019463/how-to-detect-device-name-in-safari-on-ios-13-while-it-doesnt-show-the-correct

File: src/plots/gl3d/scene.js [hunk 3]
```diff
@@ -31,6 +30,32 @@ var createAxesOptions = require('./layout/convert');
 var createSpikeOptions = require('./layout/spikes');
 var computeTickMarks = require('./layout/tick_marks');
 
+var isMobile = require('is-mobile');
+var tablet = isTablet();
+
+function isTablet() {
+    var navigator = window.navigator;
```
Comment (etpinard) [inline src/plots/gl3d/scene.js]: Does `is-mobile` setup a guard for cases when `window.navigator` is undefined?

File: src/plots/gl3d/scene.js [hunk 4]
```diff
@@ -31,6 +30,32 @@ var createAxesOptions = require('./layout/convert');
...
+    var userAgent = navigator.userAgent;
```
Comment (etpinard) [inline src/plots/gl3d/scene.js]: and similarly here, if `is-mobile` guards against undefined `navigator.userAgent`, we should too here.

File: src/plots/gl3d/scene.js (hunk 3 above)
Comment (archmoj) [reply src/plots/gl3d/scene.js]: Good call.

File: src/plots/gl3d/scene.js (hunk 4 above)
Comment (archmoj) [reply src/plots/gl3d/scene.js]: Good call.

File: src/plots/gl3d/scene.js (hunk 3 above)
Comment (archmoj) [reply src/plots/gl3d/scene.js]: I don't see any check there concerning `navigator`.

File: src/plots/gl3d/scene.js (hunk 4 above)
Comment (archmoj) [reply src/plots/gl3d/scene.js]: Revised in https://github.com/plotly/plotly.js/pull/4546/commits/436fca46dad2226913b07f3ab0d01b04cb0ad03c.

File: src/plots/gl3d/scene.js (hunk 3 above)
Comment (etpinard) [reply src/plots/gl3d/scene.js]: OK, can you add one just to be sure?

Comment (etpinard) [reply src/plots/gl3d/scene.js]: (I wouldn't someone running plotly.js in some sort of weird environment have their app break in a patch release)

Comment (archmoj) [reply src/plots/gl3d/scene.js]: Actually it is added in https://github.com/plotly/plotly.js/pull/4546/commits/436fca46dad2226913b07f3ab0d01b04cb0ad03c here:
```js
if(!ua && typeof navigator !== 'undefined') ua = navigator.userAgent;
```

Comment (etpinard) [reply src/plots/gl3d/scene.js]: Sure, we can still can breakage on this line below:

```js
navigator.maxTouchPoints > 1
```

if `window.navigator` isn't defined.

Comment (archmoj) [reply src/plots/gl3d/scene.js]: Addressed in https://github.com/plotly/plotly.js/pull/4546/commits/aa28ebafd53a0fc5cb34a54b5f0718cb0aad4f43.

**Comment by etpinard** [comment]
Sweet. Thanks very much

:dancer: - once tests pass!

**Comment by archmoj** [comment]
@etpinard Thanks for the review.
After this I would open a PR on `is-mobile` to possibly simply this for future.

**Comment by archmoj** [comment]
> @etpinard Thanks for the review.
> After this I would open a PR on `is-mobile` to possibly simply this for future.

PR to `is-mobile`: https://github.com/juliangruber/is-mobile/pull/22

**Pull request merged**